File: PROGRAMMERS/LEVEL_02/Q12949.cpp
//https://programmers.co.kr/learn/courses/30/lessons/12949
//행렬의 곱셈
#include <iostream>

#include <string>
#include <vector>

using namespace std;

vector<vector<int>> solution(vector<vector<int>> arr1, vector<vector<int>> arr2) {
    
    int Arow = arr1.size();
    int Acol = arr1[0].size(); // Acol == Brow
    int Bcol = arr2[0].size();

    // 행렬(a * k) * 행렬(k * b) --> 행렬(a * b)
    vector<vector<int>> answer(Arow, vector<int>(Bcol));
    for (int i = 0; i < Arow; ++i)
    {
        for (int j = 0; j < Bcol; ++j)
        {
            for (int k = 0; k < Acol; ++k)
            {
                answer[i][j] += arr1[i][k] * arr2[k][j];
            }
        }
    }
    return answer;
}

int main()
{
    /*
    vector<vector<int>> arr1 = { {1, 4} ,{3, 2} ,{4, 1} };
    vector<vector<int>> arr2 = { {3, 3} ,{3, 3} };
    */
    vector<vector<int>> arr1 = { {2, 3, 2} ,{4, 2, 4} ,{3, 1, 4} };
    vector<vector<int>> arr2 = { {5, 4, 3} ,{2, 4, 1} ,{3, 1, 1} };

    vector<vector<int>> answer = solution(arr1, arr2);
    for (vector<int> v : answer)
    {
        for (int i : v)
        {
            cout << i << ' ';
        }
        cout << endl;
    }

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q12949.java
// https://programmers.co.kr/learn/courses/30/lessons/12949?language=java
// 행렬의 곱셈
package PROGRAMMERS.LEVEL_02;

public class Q12949 {

    public static void main(String[] args) {
        /*
        arr1	                            arr2	                            return
        [[1, 4], [3, 2], [4, 1]]	        [[3, 3], [3, 3]]	                [[15, 15], [15, 15], [15, 15]]
        [[2, 3, 2], [4, 2, 4], [3, 1, 4]]	[[5, 4, 3], [2, 4, 1], [3, 1, 1]]	[[22, 22, 11], [36, 28, 18], [29, 20, 14]]
         */

        int[][] arr1 = {{1, 4}, {3, 2}, {4, 1}};
        int[][] arr2 = {{3, 3}, {3, 3}};
        int[][] res = solution(arr1, arr2);
        for (int[] arr : res) {
            for (int i : arr) {
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }


    public static int[][] solution(int[][] arr1, int[][] arr2) {

        int column = arr1.length;
        int row = arr2[0].length;
        int cor = arr1[0].length;
        int[][] answer = new int[column][row];

        for (int i = 0; i < column; ++i) {
            for (int j = 0; j < row; ++j) {
                for (int k = 0; k < cor; ++k) {
                    answer[i][j] += arr1[i][k] * arr2[k][j];
                }
            }
        }
        return answer;
    }
}
